### PR TITLE
cl-stitch: feature match support opencv capi

### PR DESCRIPTION
**Pull request by zongwave**

### CODE REVIEW

File: modules/ocl/cl_image_360_stitch.cpp [hunk 1]
```diff
@@ -738,8 +738,11 @@ CLImage360Stitch::sub_handler_execute_done (SmartPtr<CLImageHandler> &handler)
             convert_to_cv_rect (_img_merge_info[i].right, crop_left);
             convert_to_cv_rect (_img_merge_info[idx_next].left, crop_right);
```
Comment (windyuan) [inline modules/ocl/cl_image_360_stitch.cpp]: crop_left/crop_right is not useful any more. please remove them.

File: modules/ocl/feature_match.h [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+#include "image_file_handle.h"
+#include "ocl/cv_base_class.h"
+#include "interface/data_types.h"
+
+#include <ocl/cl_context.h>
+#include <ocl/cl_device.h>
+#include <ocl/cl_memory.h>
+
+#include <opencv2/opencv.hpp>
+#include <opencv2/core/ocl.hpp>
```
Comment (windyuan) [inline modules/ocl/feature_match.h]: Move all opencv headers to cv_feature_match and cv_capi_feature_match.
and cv_base_class should be only used in cv_feature_match.
ocl/cl_xxx.h should be removed here.

File: modules/ocl/feature_match.h [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+        : sitch_min_width (56)
+        , min_corners (8)
+        , offset_factor (0.8f)
+        , delta_mean_offset (5.0f)
+        , max_adjusted_offset (12.0f)
+    {}
+};
+
+class FeatureMatch
+    : public CVBaseClass
```
Comment (windyuan) [inline modules/ocl/feature_match.h]: No need to derive from CVBaseClass.
cv_capi_feature_match doesn't need CVBaseClass.
FeatureMatch is only like a interface for other features.
CVFeatureMatch can derive from both FeatureMatch and CVBaseClass.

File: modules/ocl/feature_match.h [hunk 4]
```diff
@@ -0,0 +1,100 @@
+/*
+ * cv_feature_match.h - optical flow feature match
```
Comment (windyuan) [inline modules/ocl/feature_match.h]: feature_match.h

Comment (zongwave) [reply modules/ocl/feature_match.h]: modified.

File: modules/ocl/feature_match.h (hunk 2 above)
Comment (zongwave) [reply modules/ocl/feature_match.h]: removed.

File: modules/ocl/feature_match.h (hunk 3 above)
Comment (zongwave) [reply modules/ocl/feature_match.h]: done.

File: modules/ocl/cl_image_360_stitch.cpp (hunk 1 above)
Comment (zongwave) [reply modules/ocl/cl_image_360_stitch.cpp]: modified function convert_to_cv_rect, please help review.

**Comment by zongwave** [comment]
update the pr according to comments, please help review, thanks.

File: modules/ocl/cv_capi_feature_match.cpp [hunk 5]
```diff
@@ -0,0 +1,201 @@
...
+}
+
+bool
+CVCapiFeatureMatch::get_crop_image (
+    SmartPtr<VideoBuffer> buffer, Rect crop_rect, std::vector<char> &crop_image, CvMat &img)
+{
+    VideoBufferInfo info = buffer->get_video_info ();
+
+    uint8_t* image_buffer = buffer->map();
+    int offset = info.width * crop_rect.pos_y + crop_rect.pos_x;
```
Comment (windyuan) [inline modules/ocl/cv_capi_feature_match.cpp]: pointer access need use stride instead of width.

File: modules/ocl/cv_capi_feature_match.cpp [hunk 6]
```diff
@@ -0,0 +1,201 @@
...
+
+    crop_image.resize (crop_rect.width * crop_rect.height);
+    for (int i = 0; i < crop_rect.height; i++) {
+        for (int j = 0; j < crop_rect.width; j++) {
+            crop_image[i * crop_rect.width + j] = image_buffer[offset + i * info.width + j];
```
Comment (windyuan) [inline modules/ocl/cv_capi_feature_match.cpp]: same as above. use stride.

**Comment by zongwave** [comment]
modified, please help review. Thanks.

File: modules/ocl/cv_capi_feature_match.cpp (hunk 6 above)
Comment (zongwave) [reply modules/ocl/cv_capi_feature_match.cpp]: fixed.

**Comment by windyuan** [comment]
looks good to me.

**Comment by windyuan** [comment]
fixed https://github.com/01org/libxcam/issues/474 and also work for AOSP.

**Pull request merged**